src/dune_file.ml

```diff
@@ -1097,15 +1097,21 @@ end
 
 module Install_conf = struct
   type file =
-    { src : string
-    ; dst : string option
+    { src : String_with_vars.t
+    ; dst : String_with_vars.t option
     }
 
   let file =
     peek_exn >>= function
-    | Atom (_, A src) -> junk >>| fun () -> { src; dst = None }
-    | List (_, [Atom (_, A src); Atom (_, A "as"); Atom (_, A dst)]) ->
-      junk >>> return { src; dst = Some dst }
+    | Atom _ | Quoted_string _ | Template _ ->
+      String_with_vars.decode >>| fun src ->
+      { src; dst = None }
+    | List (_, [_; Atom (_, A "as"); _]) ->
+      enter
+        (String_with_vars.decode >>= fun src ->
+         keyword "as" >>>
+         String_with_vars.decode >>= fun dst ->
+         return { src; dst = Some dst })
```

Comment:
In general we try to produce helpful error messages when users use a new feature without upgrading the `(lang dune ...)` line in their `dune-project` file. In this case I suggest to replace the `String_with_vars.decode` by:

```
let%map is_atom = peek_exn >>| function Atom _ -> true | _ -> false
and s = String_with_vars.decode
and version = Syntax.get_exn Stanza.syntax in
if not is_atom && version < (1, 6) then
  Syntax.Error.since (String_with_vars.loc s) Syntax.version (1, 6)
    ~what:(sprintf "Using %s here" *if String_with_vars.has_vars s then "variables" else "quoted strings");
s
```

and then bump the language version in `stanza.ml` to `(1, 6)`.

Reply:
You are right, I had seen a similar pattern in the file when I started and completely forgotten about it. Thanks